### DISCUSSION TITLE
How does the Morgan Fingerprint are folded to specific bit number?

### DISCUSSION THREAD

**Discussion by sawsimeon** [rdkit/rdkit#4398] [General]
Hi, 

I am interested to know how does the morgan fingerprint is folded to get the specific length. I have read the [article](https://pubs.acs.org/doi/10.1021/ci100050t) but fail to understand how does the fingerprint are folded into a specific size. 
This is for fingerprint size of 1000. 
```
from rdkit import Chem
from rdkit.Chem import AllChem
from rdkit import DataStructs
import numpy as np
mol = Chem.MolFromSmiles("Cc1ccccc1")
fp = AllChem.GetMorganFingerprintAsBitVect(mol, 2, nBits = 1000)
arr = np.zeros((1, ))
DataStructs.ConvertToNumpyArray(fp, arr)
len(arr)
```
And this is for fingerprint size of 60.
```
from rdkit import Chem
from rdkit.Chem import AllChem
from rdkit import DataStructs
import numpy as np
mol = Chem.MolFromSmiles("Cc1ccccc1")
fp = AllChem.GetMorganFingerprintAsBitVect(mol, 2, nBits = 60)
arr = np.zeros((1, ))
DataStructs.ConvertToNumpyArray(fp, arr)
len(arr)
```
So now the question is are the bit from fingerprint size of 60 is from the fingerprint size of 1000 or the bits are from the large  2<sup>32</sup> bitset?

**Comment by greglandrum** [comment]
It's a two step process:
1. The atom environment is hashed to create a 32bit integer. The RDKit provides a function for calculating this hash. Note that it will be different from the hash created for the same environment by other toolkits.
2. The 32bit integer is "folded" to the specifed length using the `mod` operator: `hashVal % fpSize`
